src/main/xlat_func.c

```diff
@@ -925,38 +925,76 @@ static xlat_action_t xlat_urlquote(TALLOC_CTX *ctx, fr_cursor_t *out,
  *
  * Remember to escape % with %% in strings, else xlat will try to parse it.
  */
-static ssize_t urlunquote_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
-			       UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-			       REQUEST *request, char const *fmt)
+static xlat_action_t xlat_urlunquote(TALLOC_CTX *ctx, fr_cursor_t *out,
+				   REQUEST *request, UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+				   fr_value_box_t **in)
 {
-	char const *p;
-	char *out_p = *out;
+	char const 	*p, *end;
+	char		*buff, *buff_p;
 	char *c1, *c2;
-	size_t	freespace = outlen;
+	size_t		outlen = 0;
+	fr_value_box_t	*vb;
 
-	if (outlen <= 1) return 0;
+	/*
+	 * Nothing to do if input is empty
+	 */
+	if (!(*in)) {
+		return XLAT_ACTION_DONE;
+	}
 
-	p = fmt;
-	while (*p && (--freespace > 0)) {
+	/*
+	 * Concatenate all input
+	 */
+	if (fr_value_box_list_concat(ctx, *in, in, FR_TYPE_STRING, true) < 0) {
+		RPEDEBUG("Failed concatenating input");
+		return XLAT_ACTION_FAIL;
+	}
+
+	p = (*in)->vb_strvalue;
+	end = p + (*in)->vb_length;
+
+	/*
+	 * Calculate size of output
+	 */
+	while (p < end) {
+		if (*p == '%') {
+			p += 3;
+		} else {
+			p++;
+		}
+		outlen++;
+	}
+
+	buff = buff_p = talloc_zero_array(NULL, char, outlen + 1);
```

Comment:
If it's going to be overwritten straight after allocation it's better just to use talloc_array.